Added password strength checker class and Sync listener.

zinigor opened the pull request:
This adds a class that listens to password verification events and checks the passwords for certain criteria. 

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Adds a password checker class.
* Adds a new Sync listener warning for insecure passwords.

#### Testing instructions:
* Use a Jetpack site with this PR.
* Sandbox your connection and apply D25323-code to your sandbox.
* Set an insecure password for your user, for example, the same as your login.
* Observe an audit log entry and a sticker in the MC.

#### Proposed changelog entry for your changes:
* Added a password checker class.

zinigor committed:
Added the password strength checker sync listener.

As an accompanying code, also added the password strength checker library.

zinigor committed:
Added a test for secure passwords.

enejb commented on sync/class.jetpack-sync-module-auth.php:
```diff
@@ -0,0 +1,76 @@
+			return $user;
+		}
+
+		call_user_func(
```
Instead of doing this.
I think it would be cleared to do a 
```
do_action( 'jetpack_password_check_failed', $user->ID, $test_results['test_results']['failed'] );
```

and then in listeners we can then just call 
```
add_action( 'jetpack_password_check_failed', $this->handler,  10, 2 );
```

enejb commented on tests/php/sync/test_class.jetpack-sync-module-auth.php:
```diff
@@ -0,0 +1,37 @@
+ */
+
+class WP_Test_Jetpack_Sync_Module_Auth extends WP_Test_Jetpack_Sync_Base {
+
```
❤️ Thanks for adding the test.

enejb commented:
I am not 100% since I haven't tested this but it looks like  `authenticate` filter is called on every login as well. Which is not some that we want. 

I think it would be really interesting to always send an event when the user password changes. 
and if they change the password to something that is not secure also send that as extra data with the error that we do want.

kraftbj commented:
> I am not 100% since I haven't tested this but it looks like authenticate filter is called on every login as well. Which is not some that we want.

At least use `wp_authenticate_user` instead?

> I think it would be really interesting to always send an event when the user password changes.
and if they change the password to something that is not secure also send that as extra data with the error that we do want.

Of interest:
```
/**
	 * Fires before the password and confirm password fields are checked for congruity.
	 *
	 * @since 1.5.1
	 *
	 * @param string $user_login The username.
	 * @param string $pass1     The password (passed by reference).
	 * @param string $pass2     The confirmed password (passed by reference).
	 */
	do_action_ref_array( 'check_passwords', array( $user->user_login, &$pass1, &$pass2 ) );
```

I think that may be the only place we can access a changing password before it is hashed when an user is inserted/updated.

kraftbj committed:
Add files to phpcs whitelist

kraftbj commented on _inc/lib/class.jetpack-password-checker.php:
```diff
@@ -0,0 +1,1288 @@
+<?php //phpcs:ignore WordPress.Files.FileName.InvalidClassFileName
```
FYI @zinigor I added these ignore statements to the new files and added them to the phpcs whitelist since they're new and clean.

kraftbj commented on _inc/lib/class.jetpack-password-checker.php:
```diff
@@ -0,0 +1,1288 @@
+		} else {
+			$this->user_id = $user;
+		}
+		$this->min_password_length = apply_filters( 'better_password_min_length', $this->min_password_length );
```
Needs docblock and likely changed hook name to `jetpack_password_checker`, etc

kraftbj commented on _inc/lib/class.jetpack-password-checker.php:
```diff
@@ -0,0 +1,1288 @@
+		}
+
+		/**
+		 * Filters Jetpack's password strength enforcement settings. You can modify the minimum
```
We could include in the docs that 28 is weak, 32 is suggested, etc?

kraftbj commented on sync/class.jetpack-sync-module-auth.php:
```diff
@@ -0,0 +1,76 @@
+		$this->handler = $callable;
+
+		// User authentication.
+		add_filter( 'authenticate', array( $this, 'check_password' ), 1000, 3 );
```
+1 to the non-review comments wrt using this vs other authenticate filters.

zinigor committed:
Used a compatible assertion to avoid fatals in Travis.

zinigor committed:
Merge branch 'add/password-strength-checker' of github.com:Automattic/jetpack into add/password-strength-checker

zinigor committed:
Fixed warnings to go ahead with the tests.

zinigor committed:
Moved the password checking into the User listener.

enejb committed:
Update the jetpack_wp_login event to send IP Address.

enejb committed:
Remove the additional clear_flags calls.

Only call clear_flags when we after we acually use the flags.
We had an issue where we were cleading the flags after login in before we could acually use the flags.

enejb commented:
I tested this and it works as expected for me. 

When I login as a user on my jetpack site as that has a week password. I can see the event come though on the .com side.

the code on .com will need have this PR applied to it. 
D25997-code

enejb committed:
Remove the debug call

zinigor committed:
Whitespace cleanup.
